### ISSUE TITLE
GraphQL::Query::Arguments::ArgumentValue no longer accessible via visitor.arguments_for

### ISSUE THREAD

**Issue by chrisbutcher** [rmosolgo/graphql-ruby#2841]
# Problem

I'm working on getting [graphql-metrics](https://github.com/Shopify/graphql-metrics) compatible with graphql-ruby 1.10.x, and I've encountered an issue while trying to understand how `visitor.arguments_for` works in recent versions of graphql-ruby, in particular when the new interpreter is enabled.

On 1.9.x these calls work just fine: https://github.com/Shopify/graphql-metrics/blob/master/lib/graphql/metrics/analyzer.rb#L39-L40

We opted to recurse over `arguments.argument_values.values`  (see `#extract_arguments` in the above file), in order to get metrics metadata off of instances of `GraphQL::Query::Arguments::ArgumentValue`.

As of 1.10.0, when `interpreter? => true`, `visitor.arguments_for` now returns a hash containing only the arg names pointing to scalars, rather than a number of wrapper objects which may contain instances of `GraphQL::Query::Arguments::ArgumentValue`, via `GraphQL::Execution::Interpreter::ArgumentsCache` : https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/execution/interpreter/arguments_cache.rb#L6

If we want to continue to gather metrics like `ArgumentValue#default_used?` and `#value` from arguments passed to a field visited via `GraphQL::Analysis::AST::Analyzer#on_leave_field` we need an alternate method to get `ArgumentValue`s, it seems.

I've looked at https://graphql-ruby.org/queries/interpreter.html and the 1.10.x changelog notes.

By the way, I recall that the reason we opted to stick with doing everything (including recursing through args) via `on_leave_field` is that `on_leave_arguments` skipped some things not in the AST, like arguments provided via variables.

# Possible solutions

I spoke today with @rmosolgo on the GraphQL Slack org's #ruby channel. Highlights included:

> Robert: `default_used` is basically not set by the interpreter, and it’s not provided to resolvers, since they all get plain hashes now
> [this] will require some improvements, because `default_used?` is not available at all right now 
> gotta decide whether to rework arguments_for(…) to include that info, or provide another API that can return it
> 
> Chris: makes sense! 
> Just so I can better understand: was the change in contract for `arguments_for` to return a simplified hash the result of _both_ the bump from 1.9 to 1.10 _and_ enabling the new interpreter? https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/query.rb#L248-L249
> 
> Robert: in 1.10, that method started behaving differently when `interpreter?`
> so, i think “yes” in the sense that both 1.10 and `interpreter?` are necessary causes

Robert then suggested a few ideas for a solution in a future 1.10.x release:

> I can imagine some improvement like this:

```diff
-    def arguments_for(ast_node, definition, parent_object: nil)
+    def arguments_for(ast_node, definition, parent_object: nil, as_hash: true)
       if interpreter?
         @arguments_cache ||= Execution::Interpreter::ArgumentsCache.new(self)
-        @arguments_cache.fetch(ast_node, definition, parent_object)
+        arguments = @arguments_cache.fetch(ast_node, definition, parent_object)
+        # return an `Arguments` instance if requested, otherwise return a plain Hash
+        as_hash ? arguments.to_h : arguments 
```

> or adding `.to_h` calls in places that call `.arguments_for`. (also Interpreter::ArgumentsCache would need an upgrade)
> or maybe the solution is to stop using Intepreter::ArgumentsCache, and instead do

```ruby
if interpreter && as_hash 
  arguments.to_h
else 
  arguments 
end 
```

> Chris: yeah that's interesting. the to_h struck me as something which may be good to opt-in/out
> Is there a possible performance regression if apps want that extra detail, and opt not to get the to_h result? Or is it all in-memory anyways?
>
> Robert: it’s possible but i doubt it’s very much
>
> Chris: 👍  I'd be happy to try my hand at a PR to do this
> Robert: 👍

**Comment by chrisbutcher** [comment]
Hey @rmosolgo thanks for pointing me in the direction I've gone so far.

I've stepped through `visitor.arguments_for` on 1.10.6 and made a few observations I'd like to run by you.

I think there's a bit of tension between two design goals, both of which seem good to me:
* Unifying arguments preparation from GraphQL arguments, variables into plain hashes, ready to be passed into resolvers, stripped of metadata
* AST visitors wanting to inspect metadata for each argument value, like the argument type, parent field type, parent input object type (if applicable), whether the default argument value was used.

Supporting both use cases seems ideal to me. I'm just not sure how the extra metadata preparation step could be re-introduced without a performance regression, and without reversing plans you have for 1.10.x and beyond.

I've hit a dead end in terms of carrying through the optional `as_hash` kwarg we talked about above, down from:

* https://github.com/rmosolgo/graphql-ruby/blob/v1.10.6/lib/graphql/analysis/ast/visitor.rb#L45
* To either side of this if/else: https://github.com/rmosolgo/graphql-ruby/blob/v1.10.6/lib/graphql/query.rb#L250-L252

It seems like as of 1.10.x, the `ArgumentValue` metadata we're interested in and that we normally get from [recursing over GraphQL::Query::Arguments etc.](https://github.com/Shopify/graphql-metrics/blob/035db92/lib/graphql/metrics/analyzer.rb#L86-L109) is no longer being instantiated by graphql-ruby.

It looks like even via the [legacy ArgumentsCache](https://github.com/rmosolgo/graphql-ruby/blob/v1.10.6/lib/graphql/query/arguments_cache.rb#L14), `GraphQL::Query::LiteralInput.from_arguments` no longer returns `ArgumentValue`s, with this [`if` block](https://github.com/rmosolgo/graphql-ruby/blob/v1.10.6/lib/graphql/query/literal_input.rb#L118) now being executed, and not the `argument_owner.arguments_class.new` call in the `else` block.

In short:

* It seems the gem provides is a hash of values. We need metadata; no way to get this.
* We can't use ArgumentValue; since the gem skips it:
  * When new interpreter is used
  * 1.10+ (LiteralInput#from_arguments)
* We can't optionally make `query.arguments_for` return `ArgumentValue` because they don't work with class-based schemas

I'm happy to contribute where I can, but I think there are some larger design questions I think we should discuss first. Thank you!

Relevant recent PR: https://github.com/rmosolgo/graphql-ruby/pull/2781/files

**Comment by eapache** [comment]
We ran into a variation of this just now in the 1.9 version too. If an argument whose type is an `InputObject` has a `prepare` method then `visitor.arguments_for` returns the prepared object, making it impossible to access the `ArgumentValue`s and metadata for the arguments provided on that input object.

**Comment by rmosolgo** [comment]
> `visitor.arguments_for` returns the prepared object

I'm not _exactly_ sure how to solve this, but I think once we have `Arguments` classes wrapping plain hashes, we can fix this _somehow_ 🤔
